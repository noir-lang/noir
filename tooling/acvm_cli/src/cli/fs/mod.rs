pub(super) mod inputs;
pub(super) mod witness;
