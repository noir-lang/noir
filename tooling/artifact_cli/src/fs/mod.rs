pub mod artifact;
pub mod inputs;
pub mod witness;
