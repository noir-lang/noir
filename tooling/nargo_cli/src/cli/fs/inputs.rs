use noirc_abi::{
    input_parser::{Format, InputValue},
    Abi, InputMap, MAIN_RETURN_NAME,
};
use std::{collections::BTreeMap, path::Path};

use crate::errors::FilesystemError;

/// Returns the circuit's parameters and its return value, if one exists.
/// # Examples
///
/// ```ignore
/// let (input_map, return_value): (InputMap, Option<InputValue>) =
///   read_inputs_from_file(path, "Verifier", Format::Toml, &abi)?;
/// ```
pub(crate) fn read_inputs_from_file<P: AsRef<Path>>(
    path: P,
    file_name: &str,
    format: Format,
    abi: &Abi,
) -> Result<(InputMap, Option<InputValue>), FilesystemError> {
    if abi.is_empty() {
        return Ok((BTreeMap::new(), None));
    }

    let file_path = path.as_ref().join(file_name).with_extension(format.ext());
    if !file_path.exists() {
        return Err(FilesystemError::MissingTomlFile(file_name.to_owned(), file_path));
    }

    let input_string = std::fs::read_to_string(file_path).unwrap();
    let mut input_map = format.parse(&input_string, abi)?;
    let return_value = input_map.remove(MAIN_RETURN_NAME);

    Ok((input_map, return_value))
}
