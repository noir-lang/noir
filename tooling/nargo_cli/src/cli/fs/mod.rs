use std::{
    fs::File,
    io::Write,
    path::{Path, PathBuf},
};

pub(super) mod inputs;
pub(super) mod program;
pub(super) mod witness;

pub(super) fn create_named_dir(named_dir: &Path, name: &str) -> PathBuf {
    std::fs::create_dir_all(named_dir)
        .unwrap_or_else(|_| panic!("could not create the `{name}` directory"));

    PathBuf::from(named_dir)
}

pub(super) fn write_to_file(bytes: &[u8], path: &Path) -> String {
    let display = path.display();

    let mut file = match File::create(path) {
        Err(why) => panic!("couldn't create {display}: {why}"),
        Ok(file) => file,
    };

    match file.write_all(bytes) {
        Err(why) => panic!("couldn't write to {display}: {why}"),
        Ok(_) => display.to_string(),
    }
}
