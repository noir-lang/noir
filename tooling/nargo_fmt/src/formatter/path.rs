use noirc_frontend::{
    ast::{Path, PathKind, UnresolvedType},
    token::{Keyword, Token},
};

use super::Formatter;

impl<'a> Formatter<'a> {
    pub(super) fn format_path(&mut self, path: Path) {
        self.skip_comments_and_whitespace();

        match path.kind {
            PathKind::Plain => (),
            PathKind::Crate => {
                self.write_keyword(Keyword::Crate);
                self.write_token(Token::DoubleColon);
            }
            PathKind::Dep => {
                self.write_keyword(Keyword::Dep);
                self.write_token(Token::DoubleColon);
            }
            PathKind::Super => {
                self.write_keyword(Keyword::Super);
                self.write_token(Token::DoubleColon);
            }
        }

        for (index, segment) in path.segments.into_iter().enumerate() {
            if index > 0 {
                self.write_token(Token::DoubleColon);
            }
            self.write_identifier(segment.ident);

            if let Some(generics) = segment.generics {
                self.format_turbofish(generics);
            }
        }
    }

    pub(super) fn format_turbofish(&mut self, generics: Vec<UnresolvedType>) {
        self.write_token(Token::DoubleColon);
        self.write_token(Token::Less);
        for (index, typ) in generics.into_iter().enumerate() {
            if index > 0 {
                self.write_comma();
                self.write_space();
            }
            self.format_type(typ);
        }

        // Skip trailing comma, if any
        self.skip_comments_and_whitespace();
        if self.is_at(Token::Comma) {
            self.bump();
        }

        self.write_token(Token::Greater);
    }
}
