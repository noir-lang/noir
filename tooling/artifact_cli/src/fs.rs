use std::path::Path;

use color_eyre::eyre::{self, Context};
use noirc_artifacts::{contract::ContractArtifact, program::ProgramArtifact};

/// A parsed JSON build artifact.
#[derive(Debug, Clone)]
pub enum Artifact {
    Program(ProgramArtifact),
    Contract(ContractArtifact),
}

impl Artifact {
    /// Try to parse an artifact as a binary program or a contract
    pub fn read_from_file(path: &Path) -> eyre::Result<Artifact> {
        let json = std::fs::read(path)
            .with_context(|| format!("failed to read artifact from '{}'", path.display()))?;

        let as_program = || serde_json::from_slice::<ProgramArtifact>(&json).map(Artifact::Program);
        let as_contract =
            || serde_json::from_slice::<ContractArtifact>(&json).map(Artifact::Contract);

        Ok(as_program().or_else(|e| as_contract().map_err(|_| e))?)
    }
}
