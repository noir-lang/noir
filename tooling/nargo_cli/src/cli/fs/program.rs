use std::path::{Path, PathBuf};

use nargo::artifacts::{contract::ContractArtifact, program::ProgramArtifact};
use noirc_frontend::graph::CrateName;

use crate::errors::FilesystemError;

use super::{create_named_dir, write_to_file};

pub(crate) fn save_program_to_file<P: AsRef<Path>>(
    program_artifact: &ProgramArtifact,
    crate_name: &CrateName,
    circuit_dir: P,
) -> PathBuf {
    let circuit_name: String = crate_name.into();
    save_build_artifact_to_file(program_artifact, &circuit_name, circuit_dir)
}

pub(crate) fn save_contract_to_file<P: AsRef<Path>>(
    compiled_contract: &ContractArtifact,
    circuit_name: &str,
    circuit_dir: P,
) -> PathBuf {
    save_build_artifact_to_file(compiled_contract, circuit_name, circuit_dir)
}

fn save_build_artifact_to_file<P: AsRef<Path>, T: ?Sized + serde::Serialize>(
    build_artifact: &T,
    artifact_name: &str,
    circuit_dir: P,
) -> PathBuf {
    create_named_dir(circuit_dir.as_ref(), "target");
    let circuit_path = circuit_dir.as_ref().join(artifact_name).with_extension("json");

    write_to_file(&serde_json::to_vec(build_artifact).unwrap(), &circuit_path);

    circuit_path
}

pub(crate) fn read_program_from_file<P: AsRef<Path>>(
    circuit_path: P,
) -> Result<ProgramArtifact, FilesystemError> {
    let file_path = circuit_path.as_ref().with_extension("json");

    let input_string =
        std::fs::read(&file_path).map_err(|_| FilesystemError::PathNotValid(file_path))?;
    let program = serde_json::from_slice(&input_string)
        .map_err(|err| FilesystemError::ProgramSerializationError(err.to_string()))?;

    Ok(program)
}
