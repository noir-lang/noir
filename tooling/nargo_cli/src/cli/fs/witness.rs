use std::path::{Path, PathBuf};

use acvm::acir::native_types::WitnessStack;
use nargo::constants::WITNESS_EXT;

use super::{create_named_dir, write_to_file};
use crate::errors::FilesystemError;

pub(crate) fn save_witness_to_dir<P: AsRef<Path>>(
    witness_stack: WitnessStack,
    witness_name: &str,
    witness_dir: P,
) -> Result<PathBuf, FilesystemError> {
    create_named_dir(witness_dir.as_ref(), "witness");
    let witness_path = witness_dir.as_ref().join(witness_name).with_extension(WITNESS_EXT);

    let buf: Vec<u8> = witness_stack.try_into()?;

    write_to_file(buf.as_slice(), &witness_path);

    Ok(witness_path)
}
