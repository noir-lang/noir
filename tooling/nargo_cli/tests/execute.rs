#[allow(unused_imports)]
#[cfg(test)]
mod tests {
    // Some of these imports are consumed by the injected tests
    use assert_cmd::prelude::*;
    use predicates::prelude::*;

    use std::collections::BTreeMap;
    use std::fs;
    use std::path::PathBuf;
    use std::process::Command;

    use super::*;

    test_binary::build_test_binary_once!(mock_backend, "../backend_interface/test-binaries");

    // include tests generated by `build.rs`
    include!(concat!(env!("OUT_DIR"), "/execute.rs"));
}
