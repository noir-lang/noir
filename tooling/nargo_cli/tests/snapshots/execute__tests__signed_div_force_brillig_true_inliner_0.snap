---
source: tooling/nargo_cli/tests/execute.rs
expression: artifact
---
{
  "noir_version": "",
  "hash": "15681438610761859435",
  "abi": {
    "parameters": [
      {
        "name": "ops",
        "type": {
          "kind": "array",
          "length": 15,
          "type": {
            "kind": "struct",
            "path": "SignedDivOp",
            "fields": [
              {
                "name": "lhs",
                "type": {
                  "kind": "integer",
                  "sign": "signed",
                  "width": 8
                }
              },
              {
                "name": "rhs",
                "type": {
                  "kind": "integer",
                  "sign": "signed",
                  "width": 8
                }
              },
              {
                "name": "result",
                "type": {
                  "kind": "integer",
                  "sign": "signed",
                  "width": 8
                }
              }
            ]
          }
        },
        "visibility": "private"
      }
    ],
    "return_type": null,
    "error_types": {
      "16761564377371454734": {
        "error_kind": "string",
        "string": "Array index out of bounds"
      },
      "17843811134343075018": {
        "error_kind": "string",
        "string": "Stack too deep"
      }
    }
  },
  "bytecode": "H4sIAAAAAAAA/9WbyW8TSxDG28FOPGNnIbx95e1bnjQ2eUq45QCENQl7IGEZMjaE/V0RQj4gxAFxQOICEn8rFOmyP1e6AlLcQI1k9cz0b6q+qu7p6elJSm5jm3rzK/n9si9LUP7r+jeum/Nltr2tMUBbWSyNJQMahwxo3GFAY9mAxooBjcMGNI4Y0Fg1oDExoDE1oLFmQGPdgMZRAxrHDGgcN6BxwoDGnQY0ThrQuMuAxs8MaPzcgMYvDGj80oDGrwxo/NqAxm8MaPzWgMbvDGj83oDGHwxo/NGAxt0GNP5kQOPPBjT+YkDjrwY0/mZA4+8GNP5hQOOfBjT+ZUDj3wY0/mNA41QEjW83/NBJoumDGn2wog9C9MGFPmjQBwNakKcFb1pQpgVbWhClBUda0KMFM1qQogUfWlChBQtaEKAXbnqhpRdGeiGjFx56oaAJO02IacJJEzqaMO1+86MHPj1Q6YFFDwQacGlAowGDbkjq8NShqMGmnL5x8l/5MvHlENQP8KNcIxF+B2l/Npvp2ndx9O9JXK/tB2+/WSRgM4L+bMTb2dfp2ZexOCiZk9cgux+Y/QpzAJgDCjMPzLzCHATmoMIcAuaQwhwG5rDCHAHmiMIcBeaowhwD5pjCLACzoDCLwCwqzBIwSwpzHJjjCnMCmBMKcxKYkwpzCphTCnMamNMKcwaYMwpzFpizCrMMzLLCnAPmnMKcB+a8wqwAs6Iwq8CsKswFYC4ozEVgLirMJWAuKcxlYC4rTA5MrjBXgLmiMGvArClMAUyhMC1gWgrTBqatMFeBuaow14C5pjDrwKwrzHVgrivMDWBuKMxNYG4qzC1gbinMbWBuK8wdYO4ozF1g7gomgX3nBv3sn56O++xsFqOu/3npIBb2XY7ju1US/pyI0wn/qYs5T9n4gzf0x3pkfnh/lJlOT09J1JU7m+PgukqnPw7aaJ69DzjZt4aAW4T9Jb8ftz9uzEUjtkG2a4uY+dxIx3W37twOznFeOc9V5EVdAnXlTr+f1B+XwQ/aYh0VwS/443FfDsM1fP1EwP+w8N+nO3BO5iUJ8EmAp3e6eb9f8z/qQ8/5GhdzzMmyuONK1mb7lTj2u+9Lw3Hsd++xkTj2m3LMx7ammJbhPNZV3ObxENuxIvhVsLni92sBu0OKXW4/zkcSJR/NgvWkgTjxHkhEnMznvqR76KXrzx2+u+9w/WMbx458GuBl3LSN++Mq1KXv4Rvj4f7L+a0JP3P+ONvW1stv3fXyF4qtBvXIr/sylN9aIMat8lsP8Bg36+H8Yk7rwta78sv376ee3/99+THyi/mqi3xNCNtzA8nXTLuk5EP6rAo9O6PoyTK2NynsS59VoZf13/Ml3c/3/f6o29xnJ4F7sAVXDXAfqk26c2MXbpP0E2uTVOhl/Q8hh49ErsdcuE0eb8GlAU6OQ3h/YnuOCX4ccsjn8R6VbcD8E9DwFGzE6xe9dkiEfelT1nMfLgfyhONORfDPRFypuGZuQHFhv3IBX6jRufD8iJ/f7zuf4/jLncHHNDO70VZvtXZ6ucUYnPBfEfwLf4x54HI7/6DUnskb7T15O/8vL4rptVzm3kHuahH8t2b3FtneditvNBrNImu9y/9rIzcU3qU4AAA=",
  "debug_symbols": "ldbRaoMwFIDhd8m1F56cxMS+yhjFtmkRRIu1g1F890XZWGkL5b8Rj/gh+IOemzmk3fW0bfvjcDGbj5vphn0ztUOfp9tcmN3Ydl172t5fNuVyiHG9/3Ju+mW8TM04mY1IVZjUH5azmP2x7ZLZ+HL+LExtXwqt/4Tzj0KxcFh4LCosAhYRi5oKKUtOhBPLiXLiOPGcVJwETiInvL7w+sLrC68vvL7w+sLry7v6lX0igZPISY2JLTkRTiwnyonjxHPC61te3/L6ltdXXl95feX1lddXXl95feX1lddXXl95fcfrO17fvasfwhNRThwnnpOKkTlPX83YNrsu/W7Sx2u/v1usp+9zetixz+OwT4frmJZt+3/RXt5JFYsQ1/98HsTVhXi3fviX0fs8hvzM/Nwf",
  "file_map": {
    "50": {
      "source": "struct SignedDivOp {\n    lhs: i8,\n    rhs: i8,\n    result: i8,\n}\n\nfn main(ops: [SignedDivOp; 15]) {\n    for i in 0..15 {\n        assert_eq(ops[i].lhs / ops[i].rhs, ops[i].result);\n    }\n}\n",
      "path": "/Users/asterite/Projects/noir/test_programs/execution_success/signed_div/src/main.nr"
    }
  },
  "names": [
    "main"
  ],
  "brillig_names": [
    "main"
  ]
}
